### ISSUE TITLE
Add timeout option to requests

### ISSUE THREAD

**Issue by abedra** [redis/hiredis#306]
While connectWithTimeout will guard against the use case of a slow initial handshake, there's nothing that guards against delays with requests on current connections. The example case here is if Redis spikes to 100% CPU utilization for any number of reasons (out of sockets, memory issues, etc) then anything left holding a connection will suffer without the ability to timeout. I have run into this a couple times now and would like to see it solved. I am happy to implement, but looking for at least a thumbs up on interest before I start.

**Comment by mattsta** [comment]
It's a great idea.  Looks like we could do timeout checking and break out pretty easily inside the reply gathering loops of `do { } while (reply == NULL)`.  It may require some tricks to clean up resources if we're aborting requests, but it's not impossible.

If you get a working version going we can figure how to get it added.

**Comment by abedra** [comment]
Great! After looking, there are certainly a few approaches that could be followed. We could go the same route as connect and have a `redisCommandWithTimeout` that accepts the timeout value (thinking that is the most appropriate high level thing to do). Beyond that is how the timeout actually works. Since this isn't on a file descriptor, select/poll are probably out of the equation. Using a signal is also out since EINTR is disregarded. The option of spawning a thread and doing a lock with a timeout is of course an option but would carry additional consequences. If we did a `do { } while (reply == NULL)` how do you suggest working around the blocking call issue?

**Comment by abedra** [comment]
I assume you are talking about isolating this snippet https://github.com/redis/hiredis/blob/master/hiredis.c#L844-L850

**Comment by mattsta** [comment]
Yup, all good points.  Individual requests are handled by the internal async framework, so we don't have a single fd to timeout against.

I was thinking pretty simple.  If there's a timeout specified, manually check for time passing until the timeout threshold is reached.  That obviously only works for blocking reads (default use case).

Integration with other event loops using the async API is a bit tricker and could require some hiredis API churn to adapt the timeout semantics all the way down to the registered event loop libraries.  Each event loop library has some sort of built-in "schedule an async timeout to be delivered later" capability that could be registered with hiredis too.

The basic/dumb/simple case probably looks like:
```erlang
commandWithTimeout(Command, Timeout) ->
  StartTime = localtime(),
  MaxTime = StartTime + Timeout,
  while (MaxTime - localtime() > 0)
    do reply stuff
```

The other issue will be which timeout functions to use.  `clock_gettime()` with `CLOCK_MONOTONIC ` is preferred because it's better on FreeBSD and Linux since it ignores physical time going backwards, but it's not supported everywhere.  `gettimeofday()` has a performance-optimized implementation in glibc on modern Linux (doesn't require switching to kernel mode) and will be supported everywhere else, but causes a performance problem on FreeBSD (global core synchronization).

**Comment by abedra** [comment]
Yeah, I'm not personally interested in dealing with the async cases, just the blocking ones. I'll do some testing to ensure that the while loop doesn't get blocked by the blocking call (rendering the timeout ineffective) and try that route. As for the timeout function, just let me know which you prefer. I am running on Linux and don't have freebsd setups available for testing.

**Comment by abedra** [comment]
Ok, spent some time investigating today. Using the following example code to run my tests. https://gist.github.com/abedra/e204904d3285cd023799. This simulates the issue without having to try and put Redis in an odd state. Will attempt various hiredis fixes based on this.

**Comment by abedra** [comment]
It turns out the clocks are completely unnecessary here. You cannot place a timer around the higher level read because `redisBufferRead` makes a `read()` call on the socket. The good news is that it does so directly on the file descriptor of the context. I just wrapped it in a `select()` and it seems to be working now. The resulting PoC looks like this https://gist.github.com/abedra/e204904d3285cd023799#file-hiredis-c. Obviously I have cleanup to do here, but my last question is propagation of the timeout to `redisBufferRead`. Would you rather have that as an argument to the function, passing in NULL when not necessary, or have another function named `redisBufferReadWithTimeout` that explicitly takes this?

**Comment by mattsta** [comment]
> You cannot place a timer around the higher level read because redisBufferRead makes a read()

That does make sense!  :)

> I just wrapped it in a select() and it seems to be working now.

For consistency with existing hiredis code, it's probably better to use `poll()` instead for the same effect.

> have another function named `redisBufferReadWithTimeout` that explicitly takes this?

It's better to add a new API function for users.  Typical "adding a new function with more options" pattern is pretty easy: rename existing function from `thing()` to `_thing(more arguments)`, then make the API functions wrappers to pass the correct arguments to the implementation function.


Thanks for figuring all of this out!

**Comment by abedra** [comment]
Sounds good to me. I'll clean it up, use `poll` over `select`, and add the new API function with fixes down the call stack. I'll send a note with a link to my branch before sending the PR so you can take one last look at the result.

**Comment by abedra** [comment]
Ok here's the raw working commit, https://github.com/repsheet/hiredis/commit/598da81c467a480cb434a2a033dae809fedffb21

This does not function combining. I took a few stabs at combining but ended up breaking backward compatibility. I did make the switch to `poll` though, so all is good there. Can you take a look and comment on how you think we should proceed? The only other question I have is should a timeout set an error on the context? A timeout is an error, but not a persistent error. Setting the error sets the context itself in a error state. A user would have to clear that if they wish to continue using the context going forward.

**Comment by mattsta** [comment]
To avoid code duplication, how about if `redisBufferRead()` just calls `redisBufferReadWithTimeout()` with a timeout of 0 to ignore the poll, so it falls through and gets the old behavior of immediately blocking on `read()`?  The same could potentially be done with `redisGetReplyWithTimeout()`.  We should avoid copy/paste of entire functions as much as possible.  :cat:

Looks good overall (though, tabs are the worst thing ever.  worse than gluten and measles *combined*).

Could you think up some timeout-related tests to add to `test.c` for each of the new API functions?  Redis has a `DEBUG SLEEP [SECONDS]` command that _may_ be able to help test reads that don't get replies before a timeout.

As far as reporting a user-requested timeout error, there's two answers: it *should* return an error, because we didn't process any data, but that wouldn't be good for users right now.  The error handling in hiredis needs some improvements because, as you noted, the entire context will go into error state and there's currently no "remove error" operation without killing the entire state.  So... for ease-of-use, until we have better error handling, we should probably _not_ set an error on user-requested timeout right now.  (but we should still test for poll returning 1 versus -1 versus 0 and return an error _if_ poll does error out)

Good job!  Almost done!

**Comment by mattsta** [comment]
Still thinking about this...

Is there a reason `redisSetTimeout(context, timeout)` doesn't work for this problem ("anything left holding a connection will suffer without the ability to timeout")?

The semantics for that timeout is:
>  this timer is restarted each time additional data are received by the protocol, and thus the limit is in effect an inactivity timer

**Comment by abedra** [comment]
I tried that at first but it didn't do what I was hoping. I'm almost done with the changes. Just have to write some tests. I have no idea how the gluten measles even made it into the code, but they will certainly be gone in the PR

**Comment by mattsta** [comment]
> I tried that at first but it didn't do what I was hoping.

Yeah, plus that function _will_ force the context into an error state on timeout, which isn't needed if we just want to re-try.


> I have no idea how the gluten measles even made it into the code, but they will certainly be gone in the PR

If only we could vaccinate people against tabs.  :bookmark_tabs:

**Comment by abedra** [comment]
Ok, the PR is ready to go minus one issue. I have two new tests, one that tests that values are returned when the timeout is not reached, and one that ensures no reply when the timeout is exceeded. The issue is that `DEBUG SLEEP` is itself a blocking call. If you issue it on the test context then issue the command, the test just sleeps then executes the command that should fail. I tried creating a separate non blocking context and issuing the sleep there, but it doesn't seem to help. Here's what the tests look like:

```c
static void test_blocking_connection_timeouts(struct config config) {
    redisContext *c, *n;
    redisReply *reply;

    c = connect(config);
    n = redisConnectNonBlock(config.tcp.host, config.tcp.port);

    test("Successfully completes a command when the timeout is not exceeded: ");
    redisCommand(c,"SET foo fast");
    reply = redisCommandWithTimeout(c, 10, "GET foo");
    test_cond(reply->type == REDIS_REPLY_STRING && memcmp(reply->str, "fast", 4) == 0);
    freeReplyObject(reply);

    test("Does not return a reply when the command times out: ");
    redisCommand(c,"SET foo fast");
    redisCommand(n, "DEBUG SLEEP 5");
    reply = redisCommandWithTimeout(c, 10, "GET foo");
    test_cond(reply == NULL);

    redisFree(n);
    disconnect(c, 0);
}
```

**Comment by mattsta** [comment]
> I tried creating a separate non blocking context and issuing the sleep there, but it doesn't seem to help.

Odd, I would think running it async would work since it will stall the _entire_ server and you can run another blocking call.

Well, you'd have to run the async command then wait until right _after_ the command is sent to the server (not just testing immediately after the async call returns).  So you could do something like: send async command (debug sleep 20), sleep locally for 1 second, then try local blocking operation with timeout.

**Comment by abedra** [comment]
Even with the sleeps it still isn't working properly.

**Comment by abedra** [comment]
Ok here's where things are at right now https://github.com/repsheet/hiredis/commit/cdf1a9acc51471cc86f769f86065fa5b9715183d

**Comment by mattsta** [comment]
Notes:
  - the diff is still difficult to compare because it's converting spaces to tabs in unchanged code making non-changed lines look changed (about 100 lines!).  It's a good idea to always run `git diff` before pushing to make sure _only_ lines you changed are actually in the commit.
  - the link for the complete diff is https://github.com/repsheet/hiredis/compare/request_timeouts
  - After working with this for a few hours today, I'm afraid the current approach _may_ be unusable.

I tested this (or tried to) for a while today and the basic problem is: when using a timeout-per-command, that *does not* stop the command from returning a result later!  So, what we get is:

```haskell

Case 1 (normal):
COMMAND SENT -> WAIT FOR TIMEOUT -> GET REPLY (NO TIMEOUT) -> RETURN REPLY.  END OF REQUEST.

Case 2 (server temporarily blocked, but not gone):
COMMAND SENT -> WAIT FOR TIMEOUT -> TIMEOUT -> RETURN NULL.  [SOME TIME LATER ... REPLY RECEIVED AND ADDED TO CLIENT BUFFER.]

Case 2 (server gone, never coming back):
COMMAND SENT -> WAIT FOR TIMEOUT -> TIMEOUT -> RETURN NULL.
```

Case 1 is perfect.  It's what we expect and it's the normal flow.

Case 2 is the problem. When we **do** get a timeout, the result will go in the client result buffer like a normal command, but it will be from an arbitrary _previous_ command, completely de-coupled from the originating command (e.g. `SET abc def`; `GET abc` -> [timeout]; `INCR foo` -> `def`).

It looks like the only way to implement this functionality is to force the timeout on the _writer_, not the reader.  Does that approach still fit your use case?  It's essentially the same format as this current attempt, but with the poll around `redisBufferWrite` functions instead of `redisBufferRead`.

Alternatively, we _can_ force the context into `err` when a timeout occurs.  That will stop the blast-from-replies-past problem because the entire connection will be unusable for anything else.  That approach just requires a complete reconnect when any user-requested timeout happens.

**Comment by duckie** [comment]
I stumbled upon this exact problem when trying to implement a timeout with the asyncrhonous API. I cant see how Case 3 could save it : the write may succeed in time, but the answer can still be stuck into the network.

I ended up writing a wrapper to wait for requests indefinitely, but call back the user when the answer times out. When the answer finally arrives, the result is dropped and the following requests can go on without decoupling them from their respective answers.

In the synchronous case, though, I would elect to consider the context as no longer usable and force it into `err`.

**Comment by abedra** [comment]
Sorry for my delayed response on this thread. I'm going to give a few things a try this week and see what seems to fit best. If putting the context in an err state is the right path, would you support some reworking of the error handling or introduction of error clearing by the client?

**Comment by abedra** [comment]
After looking at this, I think the cleanest way is to set the error on the context and reconnect. I don't necessarily like this, but with the current design of the library I think it makes the most sense. I propose doing that and creating a `reconnect` function. For ease of use purposes, I would like to add the configuration values to the context. This way reconnect can just accept the context, and renegotiate based on the values already present in the context. This would save potential context switches and page faults while that data is looked up again later in the execution of the program.

Let me know what you think of this. I'll start working on things once I hear back.

**Comment by abedra** [comment]
Bump.

**Comment by zodvik** [comment]
bump.

I want this feature, my use case being to guard against commands that do a large scan, for e.g. `keys`. If someone runs `keys` on a huge key space, sentinels will bring down that master. 

Ideally, I would set all my request timeout < `sentinel down-after-milliseconds`. That way, when I run a dangerous command, I would have explicitly change my request timeout.

**Comment by badboy** [comment]
It's on my list. I'll try to work on hiredis this week.

**Comment by abedra** [comment]
No need to do any work. I've already got a complete patch. I'm waiting for a reply to my last question

**Comment by badboy** [comment]
I didn't read the full discussion yet and mattsta is not available yet, so yes, "work" is necessary.

**Comment by abedra** [comment]
Ok, not entirely sure what you mean, but I was referring to writing code. If @mattsta is fine with my last proposal, I will simply send a PR. I already made the changes.

**Comment by badboy** [comment]
Hehe, no problem, we get the confusion sorted out. Please send the PR, I'll get back in contact with  @mattsta how to proceed

**Comment by badboy** [comment]
@zodvik The suggested implementation here probably won't mitigate the scenario you described. The `keys` command will still run on the server and will block following commands for the duration it takes to scan the dataset. Only your client side will see the timeout and will force an error state, but if you reconnect immediately the server might not be ready yet.

@abedra I added your commit in a branch, added the force-error-state and fixed the test

**Comment by abedra** [comment]
@badboy that's a start to the solution, but only part. Since it forces an error state, we need to introduce a reconnect function that makes this easier. I proposed some changes above that will make it easier. I was traveling the last few days and haven't had a chance to push things up. I'll send a PR with all of those proposed changes later today.

**Comment by badboy** [comment]
@abedra ping. Any work already done? :)

**Comment by abedra** [comment]
Yes. I realized last night when I went to send the PR that I had a couple more things left to make it uniform across the library. Adding those right now.

**Comment by abedra** [comment]
OK PR sent. Let me know where you want to go from here.

**Comment by mezzatto** [comment]
Whats wrong with redisSetTimeout()? I've been using it for years without problems...

**Comment by badboy** [comment]
I'm closing this for now in the hope `redisSetTimeout` _does_ actually solve all the problems. If not, feel free to re-open with new information.

**Comment by drknexus** [comment]
Bump/request reopen?

I'm having some trouble tracking what happened here.  Maybe @abedra and @badboy can comment?  It looks like the code referenced here, e.g. redisCommandWithTimeout worked its way out of master at some point, or never made it in, and is now isolated in the with-timeout branch that has fallen behind master?  

redisSetTimeout() may work for transactions in progress.  What I'm finding is that it doesn't work for contexts that have stalled (for unknown reasons), even if redisSetTimeout was triggered on a context created by redisConnectWithTimeout.  However, with a redis server that doesn't have a timeout, I think I'm seeing cases where a simple transaction, e.g. 'SET A 1' appears to hang/stall.  In particular, I see this error most frequently if I haven't used the context for over 40 minutes.  A redisCommandWithTimeout would seem to be exactly the creature that is missing.

**Comment by limbo127** [comment]
We are facing the same issue describe by drknexus
Regards, 
Nicolas Prochazka
